src/extension/vscode/client.ts

```diff
@@ -0,0 +1,162 @@
+    }
+  }
+
+  const ws = new WebSocket(url);
+  registerCleanup(ws.close.bind(ws, 1000));
+  ws.addEventListener("close", cleanup.bind(undefined, "WS_DISCONNECTED"), {
+    once: true,
+    signal,
+  });
+  ws.addEventListener("error", makeErrorHandler(cleanup), {
+    once: true,
+    signal,
+  });
+
+  const wsAdapter = createWsAdapter(ws, signal);
+  const wsRpcClient = createRpcClient(wsAdapter);
+  const wsRpcHandler = createRpcHandler(wsAdapter);
+  const wsActor = createActor(wsAdapter);
+  function getQueriesForClient() {
+    return getQueries(
+      getClient()?.queryManager.getObservableQueries("active") ?? new Map()
+    );
+  }
+  function getMutationsForClient() {
+    return getMutations(getClient()?.queryManager.mutationStore ?? {});
+  }
+  registerCleanup(
+    wsRpcHandler("getQueries", getQueriesForClient),
+    wsRpcHandler("getMutations", getMutationsForClient),
+    wsRpcHandler("getCache", () => getClient()?.cache.extract(true) ?? {}),
+    handleExplorerRequests(wsActor, getClient)
+  );
+
+  ws.addEventListener(
+    "open",
+    function open() {
+      const client = getClient();
+      if (!client) {
+        return;
+      }
+      wsActor.send({
+        type: "registerClient",
+        payload: {
+          id: "",
+          name:
+            "devtoolsConfig" in client ? client.devtoolsConfig.name : undefined,
+          version: client.version,
+          queryCount: getQueriesForClient().length,
+          mutationCount: getMutationsForClient().length,
+        },
+      });
+      loadErrorCodes(wsRpcClient, client.version);
+    },
+    { once: true, signal }
+  );
+
+  return {
+    connected: new Promise<void>((resolve) =>
+      ws.addEventListener("open", () => resolve(), { once: true, signal })
+    ),
+    unregister: cleanup,
+    onCleanup: registerCleanup,
+  };
+}
+
+function createWsAdapter(ws: WebSocket, signal: AbortSignal): MessageAdapter {
+  return {
+    addListener(listener) {
+      const wrappedListener = (ev: MessageEvent<any>) => {
+        listener(JSON.parse(ev.data));
+      };
+      ws.addEventListener("message", wrappedListener, { signal });
+      return () => ws.removeEventListener("message", wrappedListener);
+    },
+    postMessage(message) {
+      ws.send(JSON.stringify(message));
+    },
+  };
+}
+
+function getCleanupController() {
+  const cleanupContoller = new AbortController();
+  const cleanup = cleanupContoller.abort.bind(cleanupContoller);
+  const signal = cleanupContoller.signal;
+  setMaxListeners(20, signal);
+  function registerCleanup(
+    ...cleanupFunctions: Array<(reason: Reason) => void>
+  ) {
+    for (const cleanupFn of cleanupFunctions) {
+      if (signal.aborted) {
+        cleanupFn(signal.reason);
+      }
+      signal.addEventListener("abort", () => cleanupFn(signal.reason), {
+        once: true,
+      });
+    }
+  }
+  return {
+    cleanup: (reason: Reason = "UNREGISTER") => cleanup(reason),
+    signal,
+    registerCleanup,
+  };
```

Comment:
Since this function is called `getCleanupController`, I'd expect to see the `AbortController` returned as a property here. Any reason we couldn't replace both `cleanup` and `signal` with `controller`? Is the `cleanup` function purely for type safety?

```suggestion
  return {
    controller: cleanupController,
    registerCleanup,
  };
```

Reply:
The `cleanup` function exists to set a default parameter to `.abort` - I could name it `.abort`, but it doesn't really describe what it does:

It cleans up everything. Even if the connection has already been closed (so we have partially aborted, so to speak), it cleans up all the other stuff. It's not only called to abort the action, but to also clean up in the case something was already aborted externally.